File: src/impulse.cpp
//
// impulse.cpp
//
// Copyright 2008-2010 Mike Austin
// All rights reserved.
//

#include <fstream>

#include "impulse.h"
#include "core/all.h"
#include "runtime/all.h"

#include "core/value.cpp"
#include "core/frame.cpp"
#include "core/protos.cpp"

#include "parser/all.h"

#include "tests/core.h"
#include "tests/number.h"

//#include "addons/opengl/opengl.h"

using namespace std;
using namespace impulse;

int main( int argc, char* argv[] )
{
	Lexer lexer( argc > 1 ? *new fstream( argv[1] ) : cin );
	Parser parser( lexer );

	if (lexer.stream() == cin)
	{
		cout << "------------------------------------------------------------" << endl;
		cout << "Impulse 0.1.6 — Copyright 2008-2011 Mike Austin" << endl;

		cout << endl;
	
		cout << "sizeof (double)   = " << sizeof (double)   << endl;
		cout << "sizeof (Value)    = " << sizeof (Value)    << endl;
		cout << "sizeof (Frame)    = " << sizeof (Frame)    << endl;
		cout << "sizeof (SymbolId) = " << sizeof (SymbolId) << endl;
		cout << "sizeof (SlotMap)  = " << sizeof (SlotMap)  << endl;
		cout << "sizeof (Array)    = " << sizeof (Array)    << endl;

		cout << endl;
	}
	
	Object::instance().initSlots();
	GCArray::instance().initSlots();
	Boolean::instance().initSlots();
	Number::instance().initSlots();
	String::instance().initSlots();
	Range::instance().initSlots();
	Block::instance().initSlots();
	Method::instance().initSlots();
	Future::instance().initSlots();

	//CoreTest().runTest();
	//NumberTest().runTest();

	Value lobby = Lobby::instance();
	Value locals = *new Frame( lobby.getFrame() );

	//lobby.setSlot( Symbol::at( "gl" ), OpenGL::instance() );

	Expression expr;
	static Array args;
	args.push_back( lobby );

	do
	{
		try
		{
			if (lexer.stream() == cin) cout << "] ";
			expr = parser.parseLine();

			Value result = expr.eval( locals, args, locals );

			if (&result.getFrame() != &Void::instance())
			{
				if (lexer.stream() == cin) cout << "= ";
				if (lexer.stream() == cin) cout << result.inspect() << endl;
			}
		}
		catch( string& error )
		{
			cout << error << endl;
		}
		
		vector<Frame*>::iterator iter = Frame::_releasePool.begin();
		while (iter != Frame::_releasePool.end())
		{
			(*iter)->decRef();
				
			iter = Frame::_releasePool.erase( iter );
		}
	}
	while (!exitMainLoop);

	return 0;
}

File: ruby/impulse.rb
#!/usr/bin/env ruby
#
# impulse.rb
#
# Copyright 2010 Mike Austin
# Provided under MIT licence
#

$VERBOSE = nil

require 'runtime'
require 'parser'

class Symbol
  def to_proc
    Proc.new { |*args| self.send(self, *args) }
  end
end

include Impulse

# ---------------------------------------------------------------------------- #

puts "Ruby-Impulse 0.1.  Copyright 2010 Mike Austin"

lobby = LobbyProto.instance()

=begin
code1 = [MessageProto.new(:inspect, [])]
code2 = [MessageProto.new(:x, []), MessageProto.new(:pow_, [MessageProto.new(:x, [])])]
#code3 = [RangeProto.new(1, 2), MessageProto.new(:each, [Block.new(Expression.new(code1))])]


puts Expression.new(code1).eval.call(lobby, [], lobby)
puts Expression.new(code2).eval.call(lobby, [], lobby)

#Expression.new(code3).eval.call(lobby, [], lobby)
=end

stream = ARGV.size > 0 ? File.open(ARGV[0]) : STDIN
parser = Parser.new(Lexer.new(stream))

print "] " if parser.tty?
while expr = parser.expression()
#  begin
    result = expr.eval.call(lobby, [], lobby)
    if !result.is_a? VoidProto
      print "> " + result.to_s + "\n"
    end
#  rescue StandardError => error
#    puts error
#  end
  print "] " if parser.tty?
end
